### PR TITLE
exclude_from_weight_decay for AdamW and SGDW

### PR DESCRIPTION
**Pull request by leondgarse**
# Description
Brief Description of the PR:
- Add `exclude_from_weight_decay` for `DecoupledWeightDecayExtension` optimizers including `AdamW` and `SGDW`, like `LAMB`. There are several issues on this, like [Support exclude_from_weight_decay in AdamW #1903](https://github.com/tensorflow/addons/issues/1903) and [Add decay_var_list as init option to DecoupledWeightDecayExtension #2018](https://github.com/tensorflow/addons/issues/2018).
- This PR is a mimic of current `LAMB exclude_from_weight_decay` behavior, and it has a conflict with current `DecoupledWeightDecayExtension` `_decay_var_list`. I'm actually looking forward if a better solution can be introduced.

Fixes # (issue)
- [Support exclude_from_weight_decay in AdamW #1903](https://github.com/tensorflow/addons/issues/1903).
- [Add decay_var_list as init option to DecoupledWeightDecayExtension #2018](https://github.com/tensorflow/addons/issues/2018).
## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?
If we open the commented print line 248, then:
```py
import tensorflow as tf
from tensorflow import keras
import tensorflow_addons as tfa
xx = tf.random.uniform((1000, 32, 32, 3))
yy = tf.one_hot(tf.cast(tf.random.uniform((1000,)) * 10, 'int32'), depth=32)
mm = keras.models.Sequential([keras.layers.Input([32, 32, 3]), keras.layers.Flatten(), keras.layers.BatchNormalization(), keras.layers.Dense(32)])
mm.compile(optimizer=tfa.optimizers.AdamW(weight_decay=0.01, exclude_from_weight_decay=['/gamma', '/beta']), loss="categorical_crossentropy")
mm.fit(xx, yy)
# Filtered: batch_normalization/gamma
# Filtered: batch_normalization/beta
# Filtered: batch_normalization/gamma
# Filtered: batch_normalization/beta
# 32/32 [==============================] - 1s 6ms/step - loss: 8.2161

mm.save('aa.h5')
bb = keras.models.load_model('aa.h5')
print(bb.optimizer.exclude_from_weight_decay)
# ['/gamma', '/beta']
```
If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*

### CODE REVIEW

File: tensorflow_addons/optimizers/weight_decay_optimizers.py [hunk 1]
```diff
@@ -226,6 +238,25 @@ def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
                 grad, var, indices, apply_state=apply_state
             )
 
+    def _do_use_weight_decay(self, var):
+        """Whether to use L2 weight decay for `var`."""
+        if not self._decay_var_list or var.ref() in self._decay_var_list:
+            if self.exclude_from_weight_decay:
```
Comment (PhilJd) [inline tensorflow_addons/optimizers/weight_decay_optimizers.py:244]: Could we maybe factor this part out into a function in optimizer/utils.py as  `exclude_from_weight_decay(variable, exclude_regexes: List[str]) -> bool`? This would ensure consistency between LAMB and potential future optimizers with exclude lists for weight decay and reduce nesting.

Also, I'd consider it to be more intuitive to have decay_var_list having a higher priority over exclude_from_weight decay since it's a function argument (i.e., if decay_var_list is specified, always decay those independent of a match in exclude_weight_decay). WDYT?

File: tensorflow_addons/optimizers/weight_decay_optimizers.py [hunk 2]
```diff
@@ -226,6 +238,25 @@ def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
...
+                var_name = self._get_variable_name(var.name)
+                for r in self.exclude_from_weight_decay:
+                    if re.search(r, var_name) is not None:
+                        # print("Filtered:", var_name)
+                        return False
+            return True
+        return False
+
+    def _get_variable_name(self, param_name):
```
Comment (PhilJd) [inline tensorflow_addons/optimizers/weight_decay_optimizers.py:253]: Do we need this or wouldn't re.search find the substring anyway?

File: tensorflow_addons/optimizers/weight_decay_optimizers.py [hunk 3]
```diff
@@ -71,7 +72,12 @@ def __init__(self, weight_decay, *args, **kwargs):
     """
 
     @typechecked
-    def __init__(self, weight_decay: Union[FloatTensorLike, Callable], **kwargs):
+    def __init__(
+        self,
+        weight_decay: Union[FloatTensorLike, Callable],
+        exclude_from_weight_decay: Optional[List[str]] = None,
```
Comment (PhilJd) [inline tensorflow_addons/optimizers/weight_decay_optimizers.py:78]: Please add this to the Args section below.

**Review by PhilJd — CHANGES_REQUESTED** [review]
Thanks for the PR! Would you mind adding test cases for your changes to https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/tests/weight_decay_optimizers_test.py?

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 1 above)
Comment (leondgarse) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py]: - I added a function `is_variable_excluded_by_regexes(variable, exclude_regexes: List[str]) -> bool` in `optimizer/utils.py`. As it can also be used in `lamb.py _do_layer_adaptation`. I will change this name if not comfortable for you. It's mostly called as `not is_variable_excluded_by_regexes(...)`, or maybe prefer a function like `is_variable_not_excluded_by_regexes`? WDYT?
- I changed `_do_use_weight_decay` in `optimizers/weight_decay_optimizers.py` that `if self._decay_var_list and var.ref() in self._decay_var_list` will return `True`. Making `decay_var_list` higher priority than `exclude_from_weight_decay`.

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 2 above)
Comment (leondgarse) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py]: I'm not sure about this, as this is from `lamb.py`. I kept a function `get_variable_name` in `optimizers/utils.py` but not using it in `is_variable_excluded_by_regexes`. Maybe some consideration from the author? @junjiek

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 3 above)
Comment (leondgarse) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py:78]: Added.

**Comment by leondgarse** [comment]
Four test cases added:
- `test_exclude_weight_decay_adamw` for `AdamW` `exclude_from_weight_decay` param test.
- `test_var_list_with_exclude_list_adamw` for testing `AdamW` `decay_var_list` priority with `exclude_from_weight_decay`.
- `test_exclude_weight_decay_sgdw` and `test_var_list_with_exclude_list_sgdw` for same testing aim on `SGDW`.

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 1 above)
Comment (PhilJd) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py]: Thanks for factoring this out and making decay_var_list higher priority!

How about renaming  `is_variable_excluded_by_regexes` to `is_variable_matched_by_regexes(variable, regexes)`? I feel this is a bit easier to parse. I think having the not in front is fine as it keeps the function name simpler.

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 3 above)
Comment (PhilJd) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py:78]: Thanks! Could you also add a sentence that explains that decay_var_list in minimize takes priority over exclude_from_weight_decay if specified (and also add a corresponding sentence to the documentation to minimize)?

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 1 above)
Comment (leondgarse) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py]: Ya, ranamed `is_variable_excluded_by_regexes` to `is_variable_matched_by_regexes`, also its param `exclude_regexes` to just `regexes`.

File: tensorflow_addons/optimizers/weight_decay_optimizers.py (hunk 3 above)
Comment (leondgarse) [reply tensorflow_addons/optimizers/weight_decay_optimizers.py:78]: Updated `DecoupledWeightDecayExtension` `__init__`, `minimize` and `apply_gradients` docs, also `extend_with_decoupled_weight_decay` doc. Added `exclude_from_weight_decay` in `AdamW` and `SGDW` `**kwargs` doc.

**Comment by PhilJd** [comment]
Thanks for making these changes, looks good from my side! :)
I'm not a repo owner, so we'll need to wait for a LGTM from them.

**Review by PhilJd — APPROVED** [review]

**Review by seanpmorgan — APPROVED** [review]
LGTM. Thanks @leondgarse for the PR and @PhilJd for the review!
